data/pigui/libs/text.lua

```diff
@@ -0,0 +1,260 @@
+		end
+		if d < 1000*1000 then
+			return string.format("%0.2f", distance / 1000), lc.UNIT_KILOMETERS_PER_SECOND
+		end
+		return string.format("%0.2f", distance / 1000 / 1000), lc.UNIT_MILLION_METERS_PER_SECOND
+		-- no need for au/s
+	end,
+  Datetime = function(date)
+		local second, minute, hour, day, month, year = Game.GetPartsFromDateTime(date)
+		return string.format("%4i-%02i-%02i %02i:%02i:%02i", year, month, day, hour, minute, second)
+  end
+}
+
+
+
+ui.addFancyText = function(position, anchor_horizontal, anchor_vertical, data, bg_color)
+	-- always align texts at baseline
+	local spacing = 2
+	local size = Vector2(0, 0)
+	local max_offset = 0
+	for i=1,#data do
+		local item = data[i]
+		assert(item.text, "missing text for item in addFancyText")
+		assert(item.font, "missing font for item in addFancyText")
+		assert(item.color, "missing font for item in addFancyText")
+
+	  local is_icon = item.font.name ~= "icons"
+	  local s
+		local popfont
+	  if is_icon then
+			popfont = pigui:PushFont(item.font.name, item.font.size)
+			s = pigui.CalcTextSize(item.text)
+	  else
+			s = Vector2(item.font.size, item.font.size)
+	  end
+	  size.x = size.x + s.x
+	  size.x = size.x + spacing -- spacing
+	  size.y = math.max(size.y, s.y)
+	  max_offset = math.max(max_offset, item.font.offset)
+	  if is_icon then
+			if popfont then
+				pigui.PopFont()
+			end
+	  end
+	end
+	size.x = size.x - spacing -- remove last spacing
+	position = ui.calcTextAlignment(position, size, anchor_horizontal, nil)
+	if anchor_vertical == ui.anchor.top then
+	  position.y = position.y + size.y -- was max_offset, seems wrong
+	elseif anchor_vertical == ui.anchor.bottom then
+	  position.y = position.y - size.y + max_offset
+	end
+	if bg_color then
+		pigui.AddRectFilled(position - Vector2(textBackgroundMarginPixels, size.y + textBackgroundMarginPixels),
+												position + Vector2(size.x + textBackgroundMarginPixels, textBackgroundMarginPixels),
+												bg_color,
+												0,
+												0)
+	end
+	for i=1,#data do
+		local item = data[i]
+	  local is_icon = item.font.name ~= "icons"
+	  if is_icon then
+			ui.withFont(item.font.name, item.font.size, function()
+										local s = ui.addStyledText(position, ui.anchor.left, ui.anchor.baseline, item.text, item.color, item.font, item.tooltip)
+										position.x = position.x + s.x + spacing
+			end)
+	  else
+			local s = ui.addIcon(position, item.text, item.color, Vector2(item.font.size, item.font.size), ui.anchor.left, ui.anchor.bottom, item.tooltip)
+			position.x = position.x + s.x + spacing
+	  end
+	end
+	return size
+end
+
+ui.addStyledText = function(position, anchor_horizontal, anchor_vertical, text, color, font, tooltip, bg_color)
+	-- addStyledText aligns to upper left
+	local size = Vector2(0, 0)
+	ui.withFont(font.name, font.size, function()
+								size = pigui.CalcTextSize(text)
+								local vert
+								if anchor_vertical == ui.anchor.baseline then
+									vert = nil
```

Comment:
While we're at it, I'd appreciate it if you could fixup the various whitespace issues; we'd like to strive for a four-wide physical-TAB-only indentation. New scopes should introduce a single level of indentation, these alignment issues are left over from a previous contributor's emacs settings.

Reply:
That's now cleaned up and I've fixed the other files in libs/. I'm hesitant to make massive changes to views/ or modules/ due to how it could affect any in-progress development. On a glance through, I didn't see anything as bad as the above, but there's probably some spaces vs. tabs still there in places.